### PR TITLE
Improve entities options

### PR DESCRIPTION
**Pull request by simone-sanfratello**
https://github.com/platformatic/graphql-composer/issues/9

### updates

- `primaryKeys` and `foreignKyes` are now `keys` (see doc) in this way is now possible to specify different keys of the same entity in different subgraphs
- `adapter` renamed as `argsAdapter`; note default function is removed because seems unusual (to me) to have a standard one
- support lists and nulls on results

- results are merged by schema instead of by data, becuase wasn't possible to handle nulls and to match keys properly

### current limitations

- entities keys can be of one field, no composite keys

---

I've left many TODOs around

We should also improve test coverage and add more use cases
Also perfomance can be improved.

### CODE REVIEW

File: lib/query-builder.js [hunk 1]
```diff
@@ -23,34 +29,57 @@ class QueryBuilder {
...
-      return `(${mappedArgs.join(', ')})`
+  runArgsAdapter (results) {
+    const args = this.argsAdapter(results)
+    // TODO
+    // try {
+    //  args = this.argsAdapter(results)
+    // } catch (err) {
+    //   onError "error running argsAdapter" + info
+    //   throw err + cause
+    // }
```
Comment (mcollina) [inline lib/query-builder.js:72]: remove?

File: lib/query-builder.js [hunk 2]
```diff
@@ -195,18 +251,41 @@ class QueryBuilder {
...
-          set.push(keyFieldName)
-        }
+        const keyFieldName = keySelection(keyFields[i].field)
+        arrayPushUnique(set, keyFieldName)
       }
     }
 
+    // TODO if (set.length < 1) {
+    //   console.warn('something wrong here, no fields in the query selection')
+    // }
```
Comment (mcollina) [inline lib/query-builder.js:263]: removevm?

File: lib/utils.js [hunk 3]
```diff
@@ -4,4 +4,49 @@ function isObject (obj) {
...
+
+  if (Array.isArray(v)) {
+    return `[${v.map(v => toQueryArgs(v, false)).filter(a => a !== '').join(', ')}]`
+  }
+
+  if (typeof v === 'object') {
+    const o = Object.keys(v)
+      .map(k => ({ k, v: toQueryArgs(v[k], false) }))
+      .filter(({ v }) => !!v)
+      .map(({ k, v }) => `${k}: ${v}`)
```
Comment (mcollina) [inline lib/utils.js:25]: convert this to a for loop

**Review by mcollina — APPROVED** [review]
lgtm

File: README.md [hunk 4]
```diff
@@ -175,10 +181,10 @@ main()
...
-          - `adapter(partialResult)` (function, optional) - When resolving an entity across multiple subgraphs, an initial query is made to one subgraph followed by one or more followup queries to other subgraphs. The initial query must return enough information to identify the corresponding data in the other subgraphs. This function is invoked with the result of the initial query. It should return an object whose keys correspond to the `primaryKeyFields` configuration.
-          - `primaryKeyFields` (array of strings, required) - The fields used to uniquely identify objects of this type.
-          - `referenceListResolverName` (string, required) - The name of a resolver used to retrieve a list of objects by their primary keys.
+        - `entities` (object, optional) - Configuration object for working with entities in this subgraph. Each key in this object is the name of an entity data type. This is required if the subgraph contains any entities: all of them must be listed with their primary key. The values are objects with the the following schema:
```
Comment (cjihrig) [inline README.md:184]:
```suggestion
        - `entities` (object, optional) - Configuration object for working with entities in this subgraph. Each key in this object is the name of an entity data type. This is required if the subgraph contains any entities. The values are objects with the the following schema:
```

File: README.md [hunk 5]
```diff
@@ -175,10 +181,10 @@ main()
...
+          - `referenceListResolverName` (string, optional) - The name of the resolver used to retrieve a list of objects by their keys. Can be optional if the entity doesn't have a resolver because its keys are nested in another entity (see the [example](./examples/with-nested-keys.js)).
+          - `argsAdapter (partialResults)` (function, optional) - When resolving an entity across multiple subgraphs, an initial query is made to one subgraph followed by one or more followup queries to other subgraphs. The initial query must return enough information to identify the corresponding data in the other subgraphs. This function is invoked with the result of the initial query. It should return an object to be used as argument for `referenceListResolverName` query.
+          - `keys` (array of object, required) - The keys of the entity, to identify itself or refered to other entities. The key format is `[{ field, type }]`, for example `[{ field: 'id', type: 'Book' }]`. `type` can be omitted referring to the entity itself.
```
Comment (cjihrig) [inline README.md:187]: Can you explain why the user would ever need to specify the type here? Shouldn't the composer already know the type of every field? The type information is available at runtime as well.

File: examples/with-nested-keys.js [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+    service.server = fastify()
+
+    service.server.register(mercurius, { schema: service.schema, resolvers: service.resolvers, graphiql: true })
+    service.server.get('/.well-known/graphql-composition', async function (req, reply) {
+      return reply.graphql(getIntrospectionQuery())
+    })
+
+    service.host = await service.server.listen({ port: ++port })
+
+    console.log(name, service.host + '/graphiql')
```
Comment (cjihrig) [inline examples/with-nested-keys.js:26]: Is this intentional?

File: examples/with-nested-keys.js [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+    console.log(JSON.stringify(response, null, 2))
+    console.log('==========')
+  }
+}
+
+async function main () {
+  try {
+    await test()
+  } catch (err) {
+    console.dir(err, { depth: 99 })
```
Comment (cjihrig) [inline examples/with-nested-keys.js:379]: Why 99? Might as well make it `depth: null`.

File: lib/composer.js [hunk 8]
```diff
@@ -59,7 +59,8 @@ class Composer {
       validateObject(subgraph, `subgraphs[${i}]`)
 
       const {
-        entities = {},
+        name = '#' + i,
```
Comment (cjihrig) [inline lib/composer.js:62]: I don't think this is a good default name.

File: lib/composer.js [hunk 9]
```diff
@@ -87,69 +88,41 @@ class Composer {
...
-                `subgraphs[${i}].entities.${name}.foreignKeyFields[${k}]`
-              )
-            }
-          } else if (foreignKeyFields !== undefined) {
-            throw new TypeError(
-              `subgraphs[${i}].entities.${name}.foreignKeyFields must be an array`
-            )
+          if (referenceListResolverName && !argsAdapter) {
+            throw new TypeError(`subgraphs[${i}].entities.${name}.args must be provided`)
           }
```
Comment (cjihrig) [inline lib/composer.js:102]: Is `args` supposed to be `argsAdapter` here?

File: lib/query-builder.js [hunk 10]
```diff
@@ -142,51 +180,69 @@ class QueryBuilder {
...
+                  node,
+                  schemaNode,
+                  type,
+                  resolverName,
+                  argsAdapter,
+                  info: this.info,
+                  types: this.types,
+                  subgraph: nextSubgraph,
+                  keys: entityOnNextSubgraph.keys.filter(k => k.type === typeName),
+                  originKeys: keyFields.filter(k => k.type === typeName)
```
Comment (cjihrig) [inline lib/query-builder.js:218]: What if there is more than one field with the same type?

File: test/index.js [hunk 11]
```diff
@@ -470,13 +467,323 @@ test('throws if adapter returns non-object', async (t) => {
...
+          }
+        }
+      }
+    }
+
+    const router = await startRouter(t, ['authors-subgraph', 'books-subgraph', 'reviews-subgraph'], overrides, extend)
+
+    const response = await graphqlRequest(router, query)
+
+    deepEqual(calls, 0)
```
Comment (cjihrig) [inline test/index.js:700]:
```suggestion
    strictEqual(calls, 0)
```

**Review by cjihrig — COMMENTED** [review]
I have some concerns about performance and complexity, but maybe there are some conversations/context I no longer have.

File: README.md (hunk 5 above)
Comment (simone-sanfratello) [reply README.md:187]: These information are needed to merge results, not only to compose the queries, to properly merge arrays and handle nulls

```
keys: [{ field: 'id' }, { field: 'author.id', type: 'Author' }],
```

how can the use case in the example work without?

File: examples/with-nested-keys.js (hunk 6 above)
Comment (simone-sanfratello) [reply examples/with-nested-keys.js:26]: As it is an example, yes

**Comment by mcollina** [comment]
The reason for this change is to support calling a resolver for the given object with parameters.

In Platformatic DB, to get a few entities, you have to do `books(where: { id: { in: [42, 24] } }) { id, title }`. We don't want to have to add another resovler `booksWithId(ids: [42, 24])`

File: lib/query-builder.js (hunk 1 above)
Comment (simone-sanfratello) [reply lib/query-builder.js:72]: I'm going to implement

File: fixtures/reviews-subgraph.js [hunk 12]
```diff
@@ -67,21 +94,23 @@ const resolvers = {
...
-      book.reviews = book.reviews.map((rid) => {
+      book.reviews = book?.reviews.map((rid) => {
         return reviews[rid]
       })
 
       return book
     },
     async getReviewsByBookId (_, { id }) {
-      return books[id].reviews.map((rid) => {
+      return books?.[id].reviews.map((rid) => {
```
[the remark this answers is not in the document]
Comment (simone-sanfratello) [reply fixtures/reviews-subgraph.js]: Some tests need null values

File: lib/query-builder.js (hunk 10 above)
Comment (simone-sanfratello) [reply lib/query-builder.js]: as for now, nothing

> current limitations
> entities keys can be of one field, no composite keys

**Review by cjihrig — APPROVED** [review]

**Pull request merged**